### PR TITLE
Add CSV download via email link

### PR DESCRIPTION
**Pull request by theseanything**
# What
This adds the ability to download CSV of the data on the content page via an link emailed to the user.
Changes include:
- Adding CSV download link on content page
- Add Sidekiq process and worker support
- Add CSV export Sidekiq worker that generate the CSV file, uploads to S3 and emails the user.
- Add ActionMailer to mail the user
- Add download request confirmation page when user clicks on download CSV

# Why
This is to provide CSV format of the data on the content page. Previously the user could download the file by streaming the download as it was generated. However, this had issues with server timeouts and quirks in browsers behaviours around streaming files. 

This approach aims to generate the file before making it available for the user to download. The CSV is generated asynchronously as it is a long blocking process. The file is then uploaded to S3, as the file is too large to attach to the email directly. The user is then sent an email, to alert them that the data is ready to download. This should provide a more stable experience for downloading the data and be more consistent with other apps in GOV.UK that also use this workflow.

# Screenshots

## After
image

---
# Review Checklist
* [x] Changes in scope.
* [x] Added/updated unit tests.
* [x] Added/updated feature tests.
* [x] Added/updated relevant documentation.
* [x] Added to Trello card.

### CODE REVIEW

File: app/workers/csv_export_worker.rb [hunk 1]
```diff
@@ -38,7 +38,7 @@ def upload_to_s3(key, body)
       aws_secret_access_key: ENV['AWS_SECRET_ACCESS_KEY']
     )
 
-    directory = connection.directories.get(ENV['AWS_S3_BUCKET_NAME'])
+    directory = connection.directories.get(ENV['AWS_CSV_EXPORT_BUCKET_NAME'])
```
Comment (cbaines) [inline app/workers/csv_export_worker.rb]: This change, and the changes below would make a bit more sense in a different commit.

File: app/workers/csv_export_worker.rb [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 class CsvExportWorker
   include Sidekiq::Worker
+  sidekiq_options retry: 0
+  sidekiq_options queue: 'export_csv'
```
Comment (cbaines) [inline app/workers/csv_export_worker.rb]: :+1:

File: Gemfile.lock [hunk 3]
```diff
@@ -113,7 +113,7 @@ GEM
       fog-core
       nokogiri (>= 1.5.11, < 2.0.0)
     formatador (0.2.5)
-    gds-api-adapters (57.4.2)
+    gds-api-adapters (57.5.0)
```
Comment (cbaines) [inline Gemfile.lock]: As this change has hit master, you can probably just rebase this branch https://github.com/alphagov/content-data-admin/pull/430

**Comment by cbaines** [comment]
This looks pretty good, some changes maybe would be better split across different commits but that's all I've seen so far.

Given that the entire CSV file is held in memory, I'd be interested in knowing how this performs when exporting everything, especially how much memory the Sidekiq process eats.

File: app/workers/csv_export_worker.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
+require 'fog/aws'
+
+class CsvExportWorker
+  include Sidekiq::Worker
+  sidekiq_options retry: 0
+  sidekiq_options queue: 'export_csv'
```
Comment (pmanrubia) [inline app/workers/csv_export_worker.rb]: What about using `ActiveJob` API to set the queue name and retry options?:

```ruby
retry_on StandardError, attempts: 3
queue_as :default
```

File: app/workers/csv_export_worker.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    content_items = FindContent.enum(search_params)
+    ContentItemsCSVPresenter.new(
+      content_items,
+      search_params,
+      document_types,
+      organisations
+    )
+  end
+
+  def upload_to_s3(filename, body)
```
Comment (pmanrubia) [inline app/workers/csv_export_worker.rb]: Uploading a file to S3 seems like a concern that should not live in the job. Why not just extracting a class with this behaviour and reuse it from this class? Testing would also look better.

File: app/workers/csv_export_worker.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    basename = presenter.filename
+    csv_string = presenter.csv_rows
+
+    file_url = upload_to_s3("#{basename}.csv", csv_string)
+
+    # Send email with link
+    ContentCsvMailer.content_csv_email(recipient_address, file_url).deliver_now
+  end
+
+  def get_csv_presenter(search_params)
```
Comment (pmanrubia) [inline app/workers/csv_export_worker.rb]: Can we rename this method to `build_csv_presenter`? `get_XXX` feels a bit awkward to me :)

File: app/workers/csv_export_worker.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    document_types = FetchDocumentTypes.call[:document_types]
+    organisations = FetchOrganisations.call[:organisations]
+
+    content_items = FindContent.enum(search_params)
```
Comment (pmanrubia) [inline app/workers/csv_export_worker.rb]: Not for this PR, but something that I would address shortly is:

1. Add models `Organisation` and `DocumentType`
2. Add methods to return all organisations and document types with memoization
3. Remove those parameters from the presenter API

This would simplify how we build the presenter.

File: spec/workers/csv_export_worker_spec.rb [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    allow(FetchDocumentTypes).to receive(:call)
+      .and_return(document_types: document_types)
+    allow(FetchOrganisations).to receive(:call)
+      .and_return(organisations: organisations)
+    allow(FindContent).to receive(:enum)
+      .with(search_params)
+      .and_return(content_items)
+
+    Fog.mock!
```
Comment (pmanrubia) [inline spec/workers/csv_export_worker_spec.rb]: As per my comment above we would not need mock all this requests as they can be tested in a unit test.

File: spec/workers/csv_export_worker_spec.rb [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+  let(:organisations) do
+    [{ name: 'org', id: 'org-id', acronym: 'OI', },
+     { name: 'another org', id: 'another-org-id', },
+     { name: 'Users Org', id: 'users-org-id', acronym: 'UOI', }]
+  end
+
+  let(:content_items) do
+    [
+      {
```
Comment (pmanrubia) [inline spec/workers/csv_export_worker_spec.rb]: We are setting many values in the test that are not in used. Factories might help here. Have you considered using FactoryBot to simplify these tests? This could also benefit the feature test.

File: app/workers/csv_export_worker.rb (hunk 1 above)
Comment (theseanything) [reply app/workers/csv_export_worker.rb]: Separated out into separate commit

Comment (cbaines) [reply app/workers/csv_export_worker.rb]: Great, thanks, I'll take a look.

File: spec/workers/csv_export_worker_spec.rb (hunk 9 above)
Comment (theseanything) [reply spec/workers/csv_export_worker_spec.rb]: Agreed - however, I suggest this being a separate PR as we have a few other test cases where the FactoryBot could be used. It would take a bit effort to refactor this.

File: spec/workers/csv_export_worker_spec.rb (hunk 8 above)
Comment (theseanything) [reply spec/workers/csv_export_worker_spec.rb]: Still need to mock this for the worker as will ultimately call S3.

File: app/workers/csv_export_worker.rb (hunk 5 above)
Comment (theseanything) [reply app/workers/csv_export_worker.rb]: I've moved to being a concern

**Comment by theseanything** [comment]
@pmanrubia @cbaines I've addressed you comments!

File: spec/workers/csv_export_worker_spec.rb (hunk 8 above)
Comment (pmanrubia) [reply spec/workers/csv_export_worker_spec.rb]: @theseanything: Sorry if I wasn't clear on my previous message: in the unit test you could mock the method of the class that you have extracted (you own the API), considering that:

1. The unit test for the class that you have extracted would do the testing
2. The feature test would not mock anything
